Add ability to compute density of states (Lanczos), simpson integrator and new spslicer

AerysL opened the pull request:
This pull adds the ability to compute density of states of a matrix by the Lanczos algorithm (landos.c), a function to compute the integrals using simpson's rule and the trapezoidal rule, and a new interval partitioner.

The following files are edited or added:

- evsl.h: Include the new functions (LanDos, spslicer2, simpson, randn_double(generates a normally distributed vector)), in addition to minor spacing automatically done by my editor.
- landos.c: contains the code for the LanDos function
- makefile: add support for new functions and files
- spslice2.h: contains the code for the spslicer2 function
- vect.c: add randn_vector via Box-Muller transformation
- simpson.c: contains the code for simpson function
- LanDos.c: tests the landos, spslicer2 and simpson functions
- testMat.dat: used within the test

AerysL committed:
Start (currently nonfunctional) base code for landos

Based off of lanbounds.c this code will eventually compute the density of states of a matrix by the Lanczos algorithm. Currently in the process of converting a matlab script to c. Do not use, as is currently under construction.

AerysL committed:
Continue work on getting landos to be functional.

Changes include: getting structure setup correctly, use library provided functions, fix matlab indicies, correct some places where I thought a variable was a scalar and was really a vector, fixed signature (input/outputs), slightly improve documentation.
Still untested (even checking if it compiles), but significantly closer to a working state.

AerysL committed:
make landos compile

AerysL committed:
Add final functionality from matlab script (still probably does not work, but does compile)

Added setting of gamma2, and translated the final part related to generating DOS from small gaussians.

AerysL committed:
Improve comments, fix some memory leaks

AerysL committed:
fix linspace, add debug output

AerysL committed:
Start work on testing LanDos implementation

AerysL committed:
Fix some memory leaks with tests, move printing to test executable

AerysL committed:
Print debug info, increase number of steps/points.

AerysL committed:
Fix bug with copying of y to ydos, bring in test data from the matlab script, remove extraneous debug output, setup test script for matlab script data

AerysL committed:
Add randn_double which fills an array with normally distributed random numbers, and change landos.c to use it.

AerysL committed:
Fix bug which caused major output differences, added landos.o to objs part of makefile.

The bug was that because of the difference of order in multidimensional arrays, gamma2 was not using the correct indecies within the eigenvector matrix, leading to drastically incorrect output.

AerysL committed:
Change some variables to be const

AerysL committed:
Tidy up doxygen header a minimal amount

AerysL committed:
Add begining of work for spslice2 and simpson

AerysL committed:
Potentially complete simson.c and spslice2.c.

Both should work in theory, but have not been tested yet.

AerysL committed:
Change makefile to compile simpson.c and spslice2.c, fix typos within said files such that they compile.

AerysL committed:
Reformat landos

AerysL committed:
Move code that previous only ran on first iteration to be outside of loop (many
constants)

AerysL committed:
Continue work on landos.c

AerysL committed:
Change certain LanDos parameters to const

AerysL committed:
Improve consistency

Change evsl.h to include randn_double, simpson and spslicer2, minor formatting changes in landos.c, simpson.c spslice.c

AerysL committed:
Continue working on updating landos.c and associated functions to newest specification

Changes to fix various discrepencies from the source, including values during the test, and intvs type, spelling of simpson, and calculation of bb

AerysL committed:
Continue improving code to better reflect new specifications with a focus on simpson.c

AerysL committed:
Get landos.c, simpson.c and spslice2.c produce correct output

AerysL committed:
Return test vector generator (of ones) with normally distributed generator

AerysL committed:
Add division by zero prevention

AerysL committed:
Merge branch 'landos' into EVSL_Bsol

AerysL committed:
Move landos tests to their own directory for now

liruipeng reviewed (approved):
look good to me.

liruipeng commented on EVSL_1.0/SRC/landos.c:
```diff
@@ -0,0 +1,221 @@
+  const double tol = 1e-08;
+  double width = sigma * sqrt(-2 * log(tol));
+  linspace(aa, bb, npts, xdos);       // xdos = linspace(lm,lM, npts);
+  memset(y, 0, npts * sizeof(y[0]));  // y = zeros(size(xdos));
```
Calloc(y,...). So no need for memset(y,0,...)

AerysL replied:
Fixed, thanks

AerysL committed:
Remove memset as already calloc'd

AerysL commented on EVSL_1.0/SRC/landos.c:
```diff
@@ -0,0 +1,220 @@
+    int j;
+    for (j = 0; j < msteps; j++) {
+      // w = A*v
+      matvec_genev(A, &V[j * n], &V[(j + 1) * n]);
```
matvec_genev was removed in the EVS_Bsol, which I have not yet taken into account. I will have to update my code to accommodate this change.

liruipeng reviewed (changes requested):
I take my approval back... it seems that LanDOS only works for standard eig val problem, or B is factored, so you work on L^{-1}*A*L^{-T}. This might be why you call matvec_genev. Have you considered generalized eig problem without Cholesky factorization (with only B-sol)? It will need to change normal inner product to B-inner product. But anyway, the current code is in good shape and should be easy to accommodate B-Sol.

--------------
I might be wrong in my above comment. If these are only for standard eig problem, they are fine. You just to need to change  matvec_genev to matvec_A, which computes y=A*x. Yousef has matlab code for generalized problem as well. So, I thought this one is for everything.

AerysL committed:
Convert landos.c to support the evsldata format, rework landos test.

The landos function now uses matvec_a rather than the old matec_genev function. However, it still only works for the standard eig problem.

Changes to the tests involve creating a working makefile, updating to the evsldata format, and outputing to OUT rather than the directory it is in.

AerysL commented:
The Matlab script provided to me only supported the standard eig problem. As such I've just updated the code to support the evsldata format, and improved the tests; thanks for pointing out the changes required. I believe my next task will likely be to generalize it, so your comments will help me complete that.
